src/Microsoft.AspNetCore.Authentication/Internal/EventSpan.cs

```diff
@@ -0,0 +1,83 @@
+using System;
+using System.Diagnostics;
+
+namespace Microsoft.AspNetCore.Authentication.Internal
+{
+    // REVIEW: This file will be in a common shared-source package.
+    internal struct ValueStopwatch
+    {
+        private static readonly double TimestampToTicks = TimeSpan.TicksPerSecond / (double)Stopwatch.Frequency;
+
+        private bool _enabled;
+        private long _endTimestamp;
+        private long _startTimestamp;
+
+        public TimeSpan Elapsed => ComputeElapsedTime();
+        public bool Enabled => _enabled;
+
+        private ValueStopwatch(long startTimestamp)
+        {
+            _enabled = true;
+            _endTimestamp = 0;
+            _startTimestamp = startTimestamp;
+        }
+
+        public void Stop()
+        {
+            // Start timestamp can't be zero. Like... it would have to be literally the first thing executed when the machine boots to be 0.
+            if (_startTimestamp == 0)
+            {
+                throw new InvalidOperationException("An uninitialized, or 'default', ValueStopwatch cannot be used");
+            }
+
+            _endTimestamp = Stopwatch.GetTimestamp();
+            _enabled = false;
+        }
+
+        public static ValueStopwatch StartNew() => new ValueStopwatch(Stopwatch.GetTimestamp());
+
+        private TimeSpan ComputeElapsedTime()
+        {
+            // Start timestamp can't be zero. Like... it would have to be literally the first thing executed when the machine boots to be 0.
+            if (_startTimestamp == 0)
+            {
+                throw new InvalidOperationException("An uninitialized, or 'default', ValueStopwatch cannot be used");
+            }
+
+            var end = _endTimestamp == 0 ? Stopwatch.GetTimestamp() : _endTimestamp;
+            var timestampDelta = end - _startTimestamp;
+            var ticks = (long)(TimestampToTicks * timestampDelta);
+            return new TimeSpan(ticks);
+        }
+    }
+
+    internal static class EventSpan
+    {
+        public static EventSpan<T> Create<T>(T state, Action<T, TimeSpan> endAction) => new EventSpan<T>(endAction, state);
+    }
+
+    internal struct EventSpan<T>
+    {
+        private readonly Action<T, TimeSpan> _endAction;
```

Comment:
I know it's mentioned elsewhere, but this pattern can't really be optimized to remove the allocation without writing a lot of code. You need to either write a struct per 'span', or cache a delegate per span, or just don't bother with this.

Personally I, lean towards just leaving this type out and putting the `ValueStopWatch` in code.

Reply:
Yep, that's become fairly obvious from this. The amount of work to create a "clean" consumption API per event site is the same or worse so it doesn't really yield anything useful.